### PR TITLE
Allow demultiplexing of reads with dual indexes and mixed orientation

### PR DESCRIPTION
**Pull request by vaamb**
Modify the current implementation of demux_paired to allow demultiplexing of reads with dual indexes and mixed orientation, instead of throwing a ValueError.
Add an extra check to make sure no forward-reverse barcode pair has an identical reverse-forward pair.

Although I tested it manually, this commit does not add new unit tests

### CODE REVIEW

**Comment by ebolyen** [comment]
Thanks @vaamb! I've added this to our next release, and will be in touch with our review soon!

**Comment by ebolyen** [comment]
(The current test failure looks very promising, as it's a dead-man's switch expecting an error for the behavior you've added.)

**Comment by vaamb** [comment]
I tried to run the tests locally but somehow I couldn't pass them, even with the original version. 
Nevertheless, I've used this modification a couple of times in the last weeks (always with dual indexes and mixed orientation) and I did not observe any major issue

File: q2_cutadapt/_demux.py [hunk 1]
```diff
@@ -219,9 +250,8 @@ def demux_paired(seqs: MultiplexedPairedEndBarcodeInSequenceDirFmt,
                  cores: int = 1) -> \
                     (CasavaOneEightSingleLanePerSampleDirFmt,
                      MultiplexedPairedEndBarcodeInSequenceDirFmt):
-    if mixed_orientation and reverse_barcodes is not None:
```
Comment (lizgehret) [inline q2_cutadapt/_demux.py:222]: Hey @vaamb you're seeing a test failure ([here](https://github.com/qiime2/q2-cutadapt/actions/runs/4499221194/jobs/7916815093?pr=55#step:4:753)) because this error branch has been removed.

Comment (vaamb) [reply q2_cutadapt/_demux.py:222]: Thanks, I'll have a look. But first I need to go through the dev docs to be able to run the tests locally; somehow, I can’t make them work.

Comment (lizgehret) [reply q2_cutadapt/_demux.py:222]: @vaamb you should be able to run `pytest` while sitting in the top-level directory of q2-cutadapt to run these unit tests locally. Let us know if there's anything specific that you're stuck on while trying to test!

Comment (lizgehret) [reply q2_cutadapt/_demux.py:222]: @vaamb can you provide a screenshot of the entire command you're running, along with the full traceback with the test failures?

Comment (vaamb) [reply q2_cutadapt/_demux.py:222]: @lizgehret I was launching `pytest` via Pycharm and somehow it messed things up, by simply running it in the terminal I made it work!

I modified one existing test to test if some "barcode1-barcode2" and "barcode2-barcode1" pairs are present and it works as expected.

However, I have recently found an edge case ("barcode1-barcode1") that technically is correct for demultiplexing but can lead to some downstream issues (as both sequences and their reverse complement are considered as distinct). 
One solution is to allow it and advise to do a de novo clustering using vsearch with the strand parameter set to both (I made a pull request at qiime2/q2-vsearch#90 to expose it). That is what I currently implemented using `print()` (is there any other mean to log such a warning?)
The other solution would be to raise an error before

File: q2_cutadapt/tests/test_demux.py [hunk 2]
```diff
@@ -501,7 +501,7 @@ def test_di_mismatched_barcodes(self):
                                  forward_barcodes=forward_barcodes,
                                  reverse_barcodes=reverse_barcodes)
 
-    def test_di_duplicate_barcode_pairs(self):
+    def test_di_duplicate_barcode_pairs_not_mixed(self):
```
Comment (lizgehret) [inline q2_cutadapt/tests/test_demux.py:504]:
```suggestion
    def test_dual_index_duplicate_barcode_pairs_not_mixed(self):
```
This will be easier to understand for anyone who is just looking through the code in the future.

File: q2_cutadapt/tests/test_demux.py [hunk 3]
```diff
@@ -519,35 +519,24 @@ def test_di_duplicate_barcode_pairs(self):
                                  forward_barcodes=forward_barcodes,
                                  reverse_barcodes=reverse_barcodes)
 
-    def test_multiple_orientations_dual_indices(self):
+    def test_di_duplicate_barcode_pairs_mixed(self):
```
Comment (lizgehret) [inline q2_cutadapt/tests/test_demux.py:522]:
```suggestion
    def test_dual_index_duplicate_barcode_pairs_mixed(self):
```
Same as other test comment

File: q2_cutadapt/tests/test_demux.py [hunk 4]
```diff
@@ -519,35 +519,24 @@ def test_di_duplicate_barcode_pairs(self):
...
-                self.demux_paired_fn(mixed_orientation_sequences,
-                                     forward_barcodes=forward_barcodes,
-                                     reverse_barcodes=reverse_barcodes,
-                                     mixed_orientation=True)
+        with self.assertRaisesRegex(
+                ValueError, 'duplicate barcode.*sample_c.*sample_d'):
+            self.demux_paired_fn(self.muxed_sequences,
+                                 forward_barcodes=forward_barcodes,
+                                 reverse_barcodes=reverse_barcodes,
+                                 mixed_orientation=True)
```
Comment (lizgehret) [inline q2_cutadapt/tests/test_demux.py:540]: I think it would be good to add a unit test to confirm that this new behavior works as expected - so adding a test with dual-index mixed orientation reads that will run through cutadapt and pass, and possibly assert that the resulting artifact contents are as expected.

**Review by lizgehret — CHANGES_REQUESTED** [review]
Thanks for your work on this @vaamb! Overall, this generally looks good - some comments inline, mainly with some variable renaming for clarification and an additional unit test to confirm new behavior is as expected.

File: q2_cutadapt/_demux.py [hunk 5]
```diff
@@ -122,39 +122,80 @@ def _write_empty_fastq_to_mux_barcode_in_seq_fmt(seqs_dir_fmt):
...
-                             ' pairs: %s' %
-                             ', '.join(sorted(samples_w_dup_barcode_pairs)))
+    if samples_w_dup_barcode_pairs:
+        raise ValueError('The following samples have duplicate barcode '
+                         '(note: if your reads are in dual index mixed'
+                         'orientation, forward-reverse pairs are also '
+                         'used as reverse-forward pairs): %s'
+                         % ', '.join(sorted(samples_w_dup_barcode_pairs)))
+    if samples_w_identical_f_r:
+        print("The following samples are using identical barcode for forward "
```
Comment (lizgehret) [inline q2_cutadapt/_demux.py:181]: If this branch shouldn't raise an error, it would be better to add this as a warning (vs. just a print statement).

File: q2_cutadapt/_demux.py [hunk 6]
```diff
@@ -122,39 +122,80 @@ def _write_empty_fastq_to_mux_barcode_in_seq_fmt(seqs_dir_fmt):
...
-        rev_barcode_name = reverse_barcodes.name
         rev_barcodes = reverse_barcodes.to_series()
-        # 'sort = false' below prevents a warning about future behavior changes
-        # by selecting the future behavior explicitly
-        barcodes = pd.concat([barcodes, rev_barcodes], axis=1, sort=False)
-
-        for sample_id, f_barcode, r_barcode in barcodes.itertuples():
+    else:
+        # Quite ugly but it allows to continue to use the same logic as before
+        rev_barcodes = forward_barcodes.to_series()
```
[the remark this answers is not in the document]
Comment (vaamb) [reply q2_cutadapt/_demux.py:135]: As the `rev_barcodes` is not used if `reverse_barcodes` was not provided, let's say it should not be unsafe.
Alternatively, I could create a `Series` of the same length than `forward_barcodes` and filled with `None`. By doing so, I could also later use `if r_barcode is None:` rather than the current `if reverse_barcodes is None:` which could make the code clearer

File: q2_cutadapt/tests/test_demux.py (hunk 4 above)
Comment (vaamb) [reply q2_cutadapt/tests/test_demux.py:540]: I'll try to do that !

File: q2_cutadapt/_demux.py (hunk 5 above)
Comment (vaamb) [reply q2_cutadapt/_demux.py:181]: Should I use a simple `warnings.warn()` or is there a more "QIIMEesque" way to do so ?

File: q2_cutadapt/_demux.py (hunk 6 above)
[the remark this answers is not in the document]
Comment (vaamb) [reply q2_cutadapt/_demux.py:135]: In the end, I ended up branching at the very first `if reverse_barcodes is not None:` and separate both logics, which might be clearer for future readers

**Comment by vaamb** [comment]
Strange failure, it seems the new test data could not be found

**Review by lizgehret — APPROVED** [review]
These changes look great, thanks @vaamb! I'll go ahead and get this merged.

**Comment by vaamb** [comment]
@lizgehret just a small note, in order to handle the edge case when using the same barcode both for forward and reverse, I put a warning and advised to do a de novo clustering using vsearch with the strand parameter set to both, which is currently not possible. 
However, I made a pull request at https://github.com/qiime2/q2-vsearch/pull/90 to expose it. The changes are minors, I simply copied the parameters from the other clustering methods

**Comment by lizgehret** [comment]
Thanks @vaamb! Will take a look at that PR shortly - this functionality will only exist in our development environment currently, but will make sure that's addressed before our next release when this will be fully available for all users.
